File: 프로그래머스/4/42897. 도둑질/도둑질.js
// dp[k] = max(dp[k - 1], dp[k - 2] + house[k])

function solution(money) {
    const dp = Array.from({ length: money.length, }, () => 0);
    
    if (money.length === 1) {
        return money[0];
    }
    
    if (money.length <= 2) {
        return Math.max(money[0], money[1]);
    }
    
    for (let i = 0; i < dp.length - 1; i++) {
        if (i === 0) {
            dp[i] = money[i];
            continue;
        }
        if (i === 1) {
            dp[i] = Math.max(dp[i], dp[i - 1]);
            continue;
        }
        dp[i] = Math.max(dp[i - 1], dp[i - 2] + money[i]);
    }

    let max = dp[dp.length - 2];
    
    for (let i = 0; i < dp.length; i++) {
        dp[i] = 0;
    }
        
    for (let i = 1; i < dp.length; i++) {
        if (i === 1) {
            dp[i] = money[i];
            continue;
        }
        dp[i] = Math.max(dp[i - 1], dp[i - 2] + money[i]);
    }
    
    max = Math.max(max, dp[dp.length - 1]);
    
    return max;
    
}

File: 프로그래머스/4/42897. 도둑질/도둑질.py
def solution(money):
    reversed_money = list(reversed(money))
    return max(tabulation(money), tabulation(reversed_money))
    

    
def tabulation(money):
    dp = [ [ 0, 0, None, None ] for i in range(0, len(money)) ]
    for i in range(0, len(dp)):
        if i == 0:
            dp[0] = [ money[i], 0, True, False ]
        elif i == 1:
            dp[1] = [ money[i], dp[0][0], False, True ]
        else:
            # i를 쓰는 경우
            dp[i][0] = money[i] + dp[i - 1][1]
            dp[i][2] = dp[i - 1][3]
            
            # i를 안 쓰는 경우
            if dp[i - 1][0] > dp[i - 1][1]:
                dp[i][1] = dp[i - 1][0]
                dp[i][3] = dp[i - 1][2]
            else:
                dp[i][1] = dp[i - 1][1]
                dp[i][3] = dp[i - 1][3]
        
    if len(dp) > 1:
        last_index = len(dp) - 1
        if dp[last_index][2] == True:
            dp[last_index][0] -= money[0]
        if dp[last_index][3] == True:
            dp[last_index][1] -= money[1]
            
    answer = 0
    for a, b, c, d in dp:
        answer = max(a, b, answer)
        
    return answer

# 0 <= i <= 1_000_000
# 1_000_000번 연산
# dp[0] = [ money[0], 0 ]
# dp[1] = [ money[1], dp[0] ]
# dp[i] = [ money[i] + dp[i - 1][1], max(dp[i - 1][0], dp[i - 1][1]) ]

# 추가적으로, 각각이 첫번째 요소를 사용하는지 안 하는지 별도로 boolean flag를 저장
# 마지막 집을 포함하는 결과를 구할 때 첫번째 요소를 사용하는 경우, 빼버리기

